File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/MovieController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.MovieFacade;
import cz.vhromada.catalog.facade.to.MovieTO;
import cz.vhromada.validators.exceptions.RecordNotFoundException;
import cz.vhromada.validators.exceptions.ValidationException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for movies.
 *
 * @author Vladimir Hromada
 */
@Controller("movieController")
@RequestMapping("/movies")
public class MovieController extends JsonController {

	@Autowired
	@Qualifier("movieFacade")
	private MovieFacade movieFacade;

	/** Creates new data. */
	@RequestMapping(value = "/new", method = RequestMethod.GET)
	@ResponseBody
	public void newData() {
		movieFacade.newData();
	}

	/**
	 * Returns list of movies.
	 *
	 * @return list of movies
	 */
	@RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
	@ResponseBody
	public String getMovies() {
		return serialize(movieFacade.getMovies());
	}

	/**
	 * Returns movie with ID or null if there isn't such movie.
	 *
	 * @param id ID
	 * @return movie with ID or null if there isn't such movie
	 * @throws IllegalArgumentException if ID is null
	 */
	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
	@ResponseBody
	public String getMovie(@PathVariable("id") final Integer id) {
		return serialize(movieFacade.getMovie(id));
	}

	/**
	 * Adds movie. Sets new ID and position.
	 *
	 * @param movie movie
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID isn't null
	 *                                  or czech name is null
	 *                                  or czech name is empty string
	 *                                  or original name is null
	 *                                  or original name is empty string
	 *                                  or year isn't between 1940 and current year
	 *                                  or language is null
	 *                                  or subtitles are null
	 *                                  or subtitles contain null value
	 *                                  or media are null
	 *                                  or media contain null value
	 *                                  or media contain negative value
	 *                                  or URL to ČSFD page about movie is null
	 *                                  or IMDB code isn't -1 or between 1 and 9999999
	 *                                  or URL to english Wikipedia page about movie is null
	 *                                  or URL to czech Wikipedia page about movie is null
	 *                                  or path to file with movie's picture is null
	 *                                  or note is null
	 *                                  or genres are value
	 *                                  or genres contain null value
	 *                                  or genre ID is null
	 *                                  or genre name is null
	 *                                  or genre name is empty string
	 * @throws RecordNotFoundException  if genre doesn't exist in data storage
	 */
	@RequestMapping(value = "/add", method = RequestMethod.POST)
	@ResponseBody
	public void add(final String movie) {
		movieFacade.add(deserialize(movie, MovieTO.class));
	}

	/**
	 * Updates movie.
	 *
	 * @param movie new value of movie
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID is null
	 *                                  or czech name is null
	 *                                  or czech name is empty string
	 *                                  or original name is null
	 *                                  or original name is empty string
	 *                                  or year isn't between 1940 and current year
	 *                                  or language is null
	 *                                  or subtitles are null
	 *                                  or subtitles contain null value
	 *                                  or media are null
	 *                                  or media contain null value
	 *                                  or media contain negative value
	 *                                  or URL to ČSFD page about movie is null
	 *                                  or IMDB code isn't -1 or between 1 and 9999999
	 *                                  or URL to english Wikipedia page about movie is null
	 *                                  or URL to czech Wikipedia page about movie is null
	 *                                  or path to file with movie's picture is null
	 *                                  or note is null
	 *                                  or genres are value
	 *                                  or genres contain null value
	 *                                  or genre ID is null
	 *                                  or genre name is null
	 *                                  or genre name is empty string
	 * @throws RecordNotFoundException  if movie doesn't exist in data storage
	 *                                  or genre doesn't exist in data storage
	 */
	@RequestMapping(value = "/update", method = RequestMethod.POST)
	@ResponseBody
	public void update(final String movie) {
		movieFacade.update(deserialize(movie, MovieTO.class));
	}

	/**
	 * Removes movie.
	 *
	 * @param movie movie
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if movie doesn't exist in data storage
	 */
	@RequestMapping(value = "/remove", method = RequestMethod.POST)
	@ResponseBody
	public void remove(final String movie) {
		movieFacade.remove(deserialize(movie, MovieTO.class));
	}

	/**
	 * Duplicates movie.
	 *
	 * @param movie movie
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if movie doesn't exist in data storage
	 */
	@RequestMapping(value = "/duplicate", method = RequestMethod.POST)
	@ResponseBody
	public void duplicate(final String movie) {
		movieFacade.duplicate(deserialize(movie, MovieTO.class));
	}

	/**
	 * Moves movie in list one position up.
	 *
	 * @param movie movie
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID is null
	 *                                  or movie can't be moved up
	 * @throws RecordNotFoundException  if movie doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveUp", method = RequestMethod.POST)
	@ResponseBody
	public void moveUp(final String movie) {
		movieFacade.remove(deserialize(movie, MovieTO.class));
	}

	/**
	 * Moves movie in list one position down.
	 *
	 * @param movie movie
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID is null
	 *                                  or movie can't be moved down
	 * @throws RecordNotFoundException  if movie doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveDown", method = RequestMethod.POST)
	@ResponseBody
	public void moveDown(final String movie) {
		movieFacade.remove(deserialize(movie, MovieTO.class));
	}

	/**
	 * Returns true if movie exists.
	 *
	 * @param movie movie
	 * @return true if movie exists
	 * @throws IllegalArgumentException if movie is null
	 * @throws ValidationException      if ID is null
	 */
	@RequestMapping(value = "/exists", method = RequestMethod.POST)
	@ResponseBody
	public String exists(final String movie) {
		return serialize(movieFacade.exists(deserialize(movie, MovieTO.class)));
	}

	/** Updates positions. */
	@RequestMapping(value = "/updatePositions", method = RequestMethod.GET)
	@ResponseBody
	public void updatePositions() {
		movieFacade.updatePositions();
	}

	/**
	 * Returns total count of media.
	 *
	 * @return total count of media
	 */
	@RequestMapping(value = "/totalMedia", method = RequestMethod.GET)
	@ResponseBody
	public String getTotalMediaCount() {
		return serialize(movieFacade.getTotalMediaCount());
	}

	/**
	 * Returns total length of all movies.
	 *
	 * @return total length of all movies
	 */
	@RequestMapping(value = "/totalLength", method = RequestMethod.GET)
	@ResponseBody
	public String getTotalLength() {
		return serialize(movieFacade.getTotalLength());
	}

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/MovieController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Movie
import com.github.vhromada.catalog.facade.MovieFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for movies.
 *
 * @author Vladimir Hromada
 */
@RestController("movieController")
@RequestMapping("/catalog/movies")
class MovieController(
    private val movieFacade: MovieFacade
) : AbstractController() {

    /**
     * Returns list of movies.
     *
     * @return list of movies
     */
    @GetMapping
    fun getMovies(): List<Movie> {
        return processResult(result = movieFacade.getAll())!!
    }

    /**
     * Creates new data.
     */
    @PostMapping("/new")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun newData() {
        processResult(result = movieFacade.newData())
    }

    /**
     * Returns movie with ID.
     * <br></br>
     *
     * @param id ID
     * @return movie with ID
     */
    @GetMapping("/{id}")
    fun getMovie(@PathVariable("id") id: Int): Movie {
        return processResult(result = movieFacade.get(id = id))!!
    }

    /**
     * Adds movie. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Czech name is null
     *  * Czech name is empty string
     *  * Original name is null
     *  * Original name is empty string
     *  * Year isn't between 1940 and current year
     *  * Language is null
     *  * Subtitles are null
     *  * Subtitles contain null value
     *  * Media are null
     *  * Media contain null value
     *  * Length of medium is negative value
     *  * URL to ČSFD page about movie is null
     *  * IMDB code isn't -1 or between 1 and 9999999
     *  * URL to english Wikipedia page about movie is null
     *  * URL to czech Wikipedia page about movie is null
     *  * Path to file with movie's picture is null
     *  * Note is null
     *  * Genres are null
     *  * Genres contain null value
     *  * Genre ID is null
     *  * Genre name is null
     *  * Genre name is empty string
     *  * Genre doesn't exist
     *
     * @param movie movie
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@RequestBody movie: Movie) {
        processResult(result = movieFacade.add(data = movie))
    }

    /**
     * Updates movie.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Czech name is null
     *  * Czech name is empty string
     *  * Original name is null
     *  * Original name is empty string
     *  * Year isn't between 1940 and current year
     *  * Language is null
     *  * Subtitles are null
     *  * Subtitles contain null value
     *  * Media are null
     *  * Media contain null value
     *  * Length of medium is negative value
     *  * URL to ČSFD page about movie is null
     *  * IMDB code isn't -1 or between 1 and 9999999
     *  * URL to english Wikipedia page about movie is null
     *  * URL to czech Wikipedia page about movie is null
     *  * Path to file with movie's picture is null
     *  * Note is null
     *  * Genres are null
     *  * Genres contain null value
     *  * Genre ID is null
     *  * Genre name is null
     *  * Genre name is empty string
     *  * Genre doesn't exist
     *  * Movie doesn't exist in data storage
     *
     * @param movie new value of movie
     * @return validation errors
     */
    @PostMapping("/update")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun update(@RequestBody movie: Movie) {
        processResult(result = movieFacade.update(data = movie))
    }

    /**
     * Removes movie.
     * <br></br>
     * Validation errors:
     *
     *  * Movie doesn't exist in data storage
     *
     * @param id ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("id") id: Int) {
        processResult(result = movieFacade.remove(id = id))
    }

    /**
     * Duplicates movie.
     * <br></br>
     * Validation errors:
     *
     *  * Movie doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@PathVariable("id") id: Int) {
        processResult(result = movieFacade.duplicate(id = id))
    }

    /**
     * Moves movie in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Movie can't be moved up
     *  * Movie doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@PathVariable("id") id: Int) {
        processResult(result = movieFacade.moveUp(id = id))
    }

    /**
     * Moves movie in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Movie can't be moved down
     *  * Movie doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@PathVariable("id") id: Int) {
        processResult(result = movieFacade.moveDown(id = id))
    }

    /**
     * Updates positions.
     */
    @PostMapping("/updatePositions")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updatePositions() {
        processResult(result = movieFacade.updatePositions())
    }

    /**
     * Returns total count of media.
     *
     * @return total count of media
     */
    @GetMapping("/totalMedia")
    fun getTotalMediaCount(): Int {
        return processResult(result = movieFacade.getTotalMediaCount())!!
    }

    /**
     * Returns total length of all movies.
     *
     * @return total length of all movies
     */
    @GetMapping("/totalLength")
    fun getTotalLength(): Int {
        return processResult(result = movieFacade.getTotalLength())!!.length
    }

}
